### PR TITLE
Get rid of lock in ServiceContainer; refactor ServiceContainer; fix typo in string resource

### PR DESCRIPTION
**Pull request by SergeiPavlov**
Also: test for single constructor invocation for Singletons

### CODE REVIEW

File: Orm/Xtensive.Orm.Tests.Core/IoC/MainTest.cs [hunk 1]
```diff
@@ -64,6 +65,24 @@ public void Print(string msg)
...
+
+    public void Print(string msg)
+    {
+      Debug.WriteLine(msg);
+    }
+
+    [ServiceConstructor]
+    public SlowConstructorService()
+    {
+      Interlocked.Increment(ref ConstructorInvocations);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm.Tests.Core/IoC/MainTest.cs:81]:
```suggestion
      _ = Interlocked.Increment(ref ConstructorInvocations);
```

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 2]
```diff
@@ -74,36 +79,32 @@ protected override IEnumerable<object> HandleGetAll(Type serviceType)
...
+          var type = pInfos[i].ParameterType;
+          if (creating.ContainsKey((type, managedThreadId))) {
+            throw new ActivationException(Strings.ExRecursiveConstructorParameterDependencyIsDetected);
+          }
+          args[i] = Get(type);
+        }
       }
       finally {
-        creating.Remove(serviceInfo.Type);
+        creating.TryRemove(key, out _);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/IoC/ServiceContainer.cs:107]:
```suggestion
        _ = creating.TryRemove(key, out _);
```

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 3]
```diff
@@ -74,36 +79,32 @@ protected override IEnumerable<object> HandleGetAll(Type serviceType)
...
+      var key = (serviceInfo.Type, managedThreadId);
+      if (!creating.TryAdd(key, true)) {
+        throw new ActivationException(Strings.ExRecursiveConstructorParameterDependencyIsDetected);
+      }
       var args = new object[pInfos.Length];
-      creating.Add(serviceInfo.Type);
       try {
-        for (int i = 0; i < pInfos.Length; i++)
-          args[i] = Get(pInfos[i].ParameterType);
+        for (int i = 0; i < pInfos.Length; i++) {
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/IoC/ServiceContainer.cs:98]:
```suggestion
        for (var i = 0; i < pInfos.Length; i++) {
```

File: Orm/Xtensive.Orm.Tests.Core/IoC/MainTest.cs [hunk 4]
```diff
@@ -122,5 +141,22 @@ public void SelfConsumerTest()
...
       });
     }
+
+    [Test]
+    public void SingleConstructorInvocationTest()
+    {
+      var container = ServiceContainer.Create(Configuration);
+      var prevInvocations = SlowConstructorService.ConstructorInvocations;
+      var threads = new List<Thread>();
+      for (int i = 0; i < 2; ++i) {
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm.Tests.Core/IoC/MainTest.cs:151]:
```suggestion
      for (var i = 0; i < 2; ++i) {
```

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 5]
```diff
@@ -28,43 +29,47 @@ public class ServiceContainer : ServiceContainerBase
...
+    private static readonly Func<ServiceRegistration, Pair<ConstructorInfo, ParameterInfo[]>> ConstructorFactory = serviceInfo => {
+      var mappedType = serviceInfo.MappedType;
+      var ctor = (
+        from c in mappedType.GetConstructors()
+        where c.GetAttribute<ServiceConstructorAttribute>(AttributeSearchOptions.InheritNone) != null
+        select c
+        ).SingleOrDefault() ?? mappedType.GetConstructor(Array.Empty<Type>());
+      var @params = ctor?.GetParameters();
+      return new Pair<ConstructorInfo, ParameterInfo[]>(ctor, @params);
+    };
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/IoC/ServiceContainer.cs:72]: `static readonly Func<>` is more similar to function than to other king of object
It is reasonably to put it into function section

**Pull request merged**